README.md
# commandlinefu

Simple **nodejs** module to easily get commandlinefu.com snippets in JSON format.  
[![https://nodei.co/npm/commandlinefu.png?downloads=true&downloadRank=true&stars=true](https://nodei.co/npm/commandlinefu.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/commandlinefu)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![Build Status](https://travis-ci.org/TheRealBarenziah/commandlinefu.svg?branch=master)](https://travis-ci.org/TheRealBarenziah/commandlinefu)

# Install

`npm install commandlinefu`

# Use

With nodejs:

```javascript
// clfu.js
const clfu = require("commandlinefu");

clfu()
  .then((res) => console.log(res))
  .catch((e) => e);
```

`node clfu.js` output:

```javascript
{
  "id": "5427",
  "command": "curl ifconfig.me",
  "summary": "Get your external IP address",
  "votes": "262",
  "url": "http://www.commandlinefu.com/commands/view/5427/get-your-external-ip-address"
}
```

This object will be referred as `responseObject` from now on.

# Features

By default (no argument) it will returns a random `responseObject`.  
**You can pass a _string_ as argument, in which case it will return an Array[]** of responseObject{}.  
Available arguments:

- "popular" returns an Array of popular `responseObject` (which tbh doesn't change much over time)
- "search:\$var" returns an Array of `responseObject` related to `$var`

Example of dynamic search:

```javascript
const userInput = "ssh"; // here you got user input as string

clfu(`search:${userInput}`)
  .then((res) => console.log(res))
  .catch((e) => e);
/* Outputs an array like : 
[
    {
        "id": "24887",
        "command": "ssh -L8888:localhost:80 -i nov15a.pem ubuntu@123.21.167.60",
        "summary": "port forwarding",
        "votes": "1",
        "url": "http://www.commandlinefu.com/commands/view/24887/port-forwarding"
    },
    {
        "id": "24879",
        "command": "rsync -e 'ssh -i /root/my.pem' -avz /mysql/db/data_summary.* ec2-1-2-4-9.compute-1.amazonaws.com:/mysql/test/",
        "summary": "rsync using pem file",
        "votes": "0",
        "url": "http://www.commandlinefu.com/commands/view/24879/rsync-using-pem-file"
    },
    {
        "id": "24863",
        "command": "cat myFile.json | ssh root@remoteSftpServer -o \"ProxyCommand=nc.openbsd -X connect -x proxyhost:proxyport %h %p\" 'cat > myFile.json'",
        "summary": "SFTP upload through HTTPS proxy",
        "votes": "0",
        "url": "http://www.commandlinefu.com/commands/view/24863/sftp-upload-through-https-proxy"
    }
]
*/
```

# Discord bot

I originally wrote this for my [discord bot](https://github.com/TheRealBarenziah/jdr9000). If you want to use this module for your own, feel free to check my (dirty) implementation :

- [here](https://github.com/TheRealBarenziah/jdr9000/blob/master/commands/clfu.js#L9) is where I take user input & call my module to fetch stuff accordingly
- [here](https://github.com/TheRealBarenziah/jdr9000/blob/master/utils/format.js#L18) are my format functions. Note that I divide the response into several chunks to get around Discord limitations (maximum 2000 characters for a message).

# CLI

Nodejs make it easy to wrap functions into a custom CLI suiting your needs. In terminal :

```bash
mkdir myCli
cd ./myCli
npm init # you can spam 'return' key for now
npm i commandlinefu # grab this module
touch index.js
```

This is the content of your `index.js` :

```javascript
const clfu = require("commandlinefu");

const commandlinefu = async () => {
  if (!process.argv.slice(2)[0]) {
    return await clfu()
    .then(res => console.log(res.command))
    .catch(e => console.error(e));
  }
  else if (process.argv.slice(2)[0]) {
    return await clfu(process.argv.slice(2)[0])
      .then(res => console.log(res))
      .catch(e => console.error(e));
};

commandlinefu();
```

That's it! Now back into your terminal:

```bash
node index.js # calling clfu without argument
ps auxw | awk '/(apache|httpd)/{print"strace -F -p " $2}' | sh # this is 'res.command' printed in stdout
node index.js search:ssh # calling clfu with "search:ssh" argument, we get an array like we're supposed to :
[
  {
    id: '24887',
    command: 'ssh -L8888:localhost:80 -i nov15a.pem ubuntu@123.21.167.60',
    summary: 'port forwarding',
    votes: '1',
    url: 'http://www.commandlinefu.com/commands/view/24887/port-forwarding'
  },
  {
    id: '24879',
    command: "rsync -e 'ssh -i /root/my.pem' -avz /mysql/db/data_summary.* ec2-1-2-4-9.compute-1.amazonaws.com:/mysql/test/",
    summary: 'rsync using pem file',
    votes: '0',
    url: 'http://www.commandlinefu.com/commands/view/24879/rsync-using-pem-file'
  },
  {...more objects}
]
```

This is the basic principle. You can parse & do your stuff with those arrays in the `else if` block of aforementioned `index.js` file.  
If you never used `process.argv` before, [this read should get you started](https://nodejs.org/en/knowledge/command-line/how-to-parse-command-line-arguments/)

# More

If you're looking for out-of-the-box CLI instead of a reusable function, it may be worth checking [this repo](https://github.com/nire0510/clfu).  
At least it helped me decipher commandlinefu.com API endpoints, so thank you nire0510 :)

# Todo

- supporting .mjs
- supporting browsers
  [CHANGELOG](https://github.com/TheRealBarenziah/commandlinefu/blob/master/CHANGELOG.md)


CHANGELOG.md
# 0.0.1

First commit.
[todo](https://stackoverflow.com/a/8420905/11894221)


src/__tests__/clfuUsingImport.test.js
var path = require("path");

test("commandlinefu using require", async () => {

  expect(
    await commandlinefu("random").then((res) => res.length),
  ).toBe(1);


});


src/__tests__/clfuUsingImport.test.mjs
var path = require("path");

test("commandlinefu using require", async () => {
  expect(await commandlinefu("random").then((res) => res.length)).toBe(1);
});


src/__tests__/clfuUsingRequire.test.js
/**
 * @jest-environment node
 */
// bc https://github.com/axios/axios/issues/1754
const commandlinefu = require("../../lib/cjs");

const doesObjectLooksOkay = (object) => {
  const keysItShouldHave = ["id", "command", "summary", "votes", "url"];
  const keysItHave = Object.keys({ ...object });

  return keysItShouldHave.length === keysItHave.length &&
    keysItShouldHave.every((value, index) => value === keysItHave[index])
    ? true // "Souka."  https://stackoverflow.com/a/19746771/11894221
    : false;
};

test("no arg, using require", async () => {
  expect(await commandlinefu().then((res) => doesObjectLooksOkay(res))).toBe(
    true,
  );
});

test("'popular' arg, using require", async () => {
  expect(
    await commandlinefu("popular").then((res) => doesObjectLooksOkay(res[0])),
  ).toBe(true);
});

test("'search:grep', using require", async () => {
  expect(
    await commandlinefu("search:grep").then((res) =>
      doesObjectLooksOkay(res[0]),
    ),
  ).toBe(true);
});


src/index.ts
import performRequest from "./performRequests";
import IResponseObject from "./IResponseObject";

/**
 * Get snippets from commandlinefu.com in JSON format
 *
 * @param {string} arg is optional. Allowed values: "popular", "search:$var" (where $var is a variable)
 * @returns {Promise.<IResponseObject>}
 *    A promise. Access your data using `.then` as shown in [the README](https://github.com/TheRealBarenziah/commandlinefu#use)
 * @example
 *     commandlinefu()
 *       .then(res => console.log(res))
 *       .catch(e => e);
 */

const commandlinefu = async (
  arg: string | void,
): Promise<IResponseObject | IResponseObject[] | Error> => {
  if (!Boolean(arg)) {
    return await performRequest("random/json", true);
  } else if (arg === "popular") {
    return await performRequest("browse/sort-by-votes/json", false);
  } else if (String(arg).includes("search:")) {
    const parsedArg = String(arg).split("search:")[1];
    const toBase64 = Buffer.from(String(parsedArg)).toString("base64");
    const formatedArg = `matching/${parsedArg}/${toBase64}/json`;
    return await performRequest(formatedArg, false);
  } else {
    return new Error("Invalid argument. Did you pass a valid string ?");
  }
};

export default commandlinefu;


toy.js
const clfu = require("./lib/cjs");
const x = [
  {
    id: "2442",
    command: 'grep -2 -iIr "err\\|warn\\|fail\\|crit" /var/log/*',
    summary: "Get all possible problems from any log files",
    votes: "6",
    url:
      "http://www.commandlinefu.com/commands/view/2442/get-all-possible-problems-from-any-log-files",
  },
];

const doesObjectLooksOkay = (object) => {
  const keysItShouldHave = ["id", "command", "summary", "votes", "url"];
  const keysItHave = Object.keys(...object);
  console.log("keysItHave ", keysItHave);
  console.log(
    "keysItShouldHave.length === keysItHave.length ",
    keysItShouldHave.length === keysItHave.length,
  );
  console.log(
    "keysItShouldHave.every((value, index) => value === keysItHave[index]",
    keysItShouldHave.every((value, index) => value === keysItHave[index]),
  );

  return keysItShouldHave.length === keysItHave.length &&
    keysItShouldHave.every((value, index) => value === keysItHave[index])
    ? true // "Souka."  https://stackoverflow.com/a/19746771/11894221
    : false;
};

clfu("popular").then((res) => console.log(res));
